kubernetes/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -0,0 +1,222 @@
+
+    // Check for any finished shutting down and kill the pods
+    val shutdown = shutdownToPod.toVector.filter { case (e, _) => !active.contains(e) }
+    killExecutorPods(shutdown)
+
+    // Now request shutdown for the new ones.
+    // Move them from executor list to list pending shutdown
+    for ((id, pod) <- idPodPairs) {
+      try {
+        // TODO: 'ask' returns a future - can it be used to check eventual success?
+        Option(driverEndpoint).foreach(_.ask[Boolean](RemoveExecutor(id, ExecutorKilled)))
+        executorToPod -= id
+        shutdownToPod += ((id, pod))
+      } catch {
+        case e: Exception => logError(s"Error shutting down executor $id", e)
+      }
+    }
+  }
+
+  private def killExecutorPods(idPodPairs: Seq[(String, String)]) {
+    for ((id, pod) <- idPodPairs) {
+      try {
+        client.pods().inNamespace(ns).withName(pod).delete()
+        executorToPod -= id
+        shutdownToPod -= id
+      } catch {
+        case e: Exception => logError(s"Error killing executor pod $pod", e)
+      }
+    }
+  }
+
+  def getInitialTargetExecutorNumber(conf: SparkConf,
+                                     numExecutors: Int =
+                                     DEFAULT_NUMBER_EXECUTORS): Int = {
+    if (dynamicExecutors) {
+      val minNumExecutors = conf.get(DYN_ALLOCATION_MIN_EXECUTORS)
+      val initialNumExecutors =
+        Utils.getDynamicAllocationInitialExecutors(conf)
+      val maxNumExecutors = conf.get(DYN_ALLOCATION_MAX_EXECUTORS)
+      require(
+        initialNumExecutors >= minNumExecutors && initialNumExecutors <= maxNumExecutors,
+        s"initial executor number $initialNumExecutors must between min executor number " +
+          s"$minNumExecutors and max executor number $maxNumExecutors")
+
+      initialNumExecutors
+    } else {
+      conf.get(EXECUTOR_INSTANCES).getOrElse(numExecutors)
+    }
+  }
+
+  def createExecutorPod(executorNum: Int): String = {
+    // create a single k8s executor pod.
+    val labelMap = Map("type" -> "spark-executor")
+    val podName = s"$sparkExecutorName-$executorNum"
+
+    val submitArgs = mutable.ArrayBuffer.empty[String]
+
+    if (conf.getBoolean("spark.dynamicAllocation.enabled", false)) {
+      submitArgs ++= Vector(
+        "dynamic-executors")
+    }
+
+    submitArgs ++= Vector("org.apache.spark.executor.CoarseGrainedExecutorBackend",
+      "--driver-url", s"$driverURL",
+      "--executor-id", s"$executorNum",
+      "--hostname", "localhost",
+      "--app-id", "1", // TODO: change app-id per application and pass from driver.
+      "--cores", "1")
+
+    var pod = new PodBuilder()
+      .withNewMetadata()
+      .withLabels(labelMap.asJava)
+      .withName(podName)
+      .endMetadata()
+      .withNewSpec()
+      .withRestartPolicy("OnFailure")
+
+      .addNewContainer().withName("spark-executor").withImage(sparkImage)
+      .withImagePullPolicy("IfNotPresent")
+      .withCommand("/opt/executor.sh")
```

Comment:
Actually I think I was mistaken here - `spark-class` eventually calls to [this class](https://github.com/apache/spark/blob/master/launcher/src/main/java/org/apache/spark/launcher/SparkClassCommandBuilder.java#L65) which then forks the actual process running the passed-in class. We can probably then indeed take advantage of `SPARK_JAVA_OPTS` and `SPARK_EXECUTOR_MEMORY`, etc. Arguments like executor-id which are passed as command line arguments and not JVM options should probably still be set as environment variables on the container though and the Dockerfile handles formatting the command line arguments sent to `spark-class`.